pdns/dnsdist-lua-actions.cc

```diff
@@ -489,6 +489,65 @@ class LuaFFIAction: public DNSAction
   func_t d_func;
 };
 
+class LuaFFIPerThreadAction: public DNSAction
+{
+public:
+  typedef std::function<int(dnsdist_ffi_dnsquestion_t* dq)> func_t;
+
+  LuaFFIPerThreadAction(const std::string& code): d_functionCode(code), d_functionID(s_functionsCounter++)
+  {
+  }
+
+  DNSAction::Action operator()(DNSQuestion* dq, std::string* ruleresult) const override
+  {
+    try {
+      auto& state = t_perThreadStates[d_functionID];
+      if (!state.d_initialized) {
+        setupLuaFFIPerThreadContext(state.d_luaContext);
+        state.d_func = state.d_luaContext.executeCode<func_t>(d_functionCode);
+        state.d_initialized = true;
+      }
+
+      dnsdist_ffi_dnsquestion_t dqffi(dq);
+      auto ret = state.d_func(&dqffi);
+      if (ruleresult) {
+        if (dqffi.result) {
+          *ruleresult = *dqffi.result;
+        }
+        else {
+          // default to empty string
+          ruleresult->clear();
+        }
+      }
+      return static_cast<DNSAction::Action>(ret);
+    } catch (const std::exception &e) {
+      warnlog("LuaFFIPerThreadAction failed inside Lua, returning ServFail: %s", e.what());
+    } catch (...) {
+      warnlog("LuaFFIPerthreadAction failed inside Lua, returning ServFail: [unknown exception]");
+    }
+    return DNSAction::Action::ServFail;
+  }
+
+  string toString() const override
+  {
+    return "Lua FFI per-thread script";
+  }
+
+private:
+  struct PerThreadState
+  {
+    LuaContext d_luaContext;
+    func_t d_func;
+    bool d_initialized{false};
+  };
+  static uint64_t s_functionsCounter;
```

Comment:
Is this multi-thread safe? I do not know enough about the dnsdist runtime to verify that.

Reply:
It should be, the actions are created one by one during the initial parsing of the configuration, and later the console user needs to hold the `g_luamutex` mutex while inserting a new rule. Still there is no good reason not to make that counter atomic since we don't care about performance, so I'll do that.

Reply:
Related question: what is the life cycle of the created Lua states? Do I understand correctly that one is created globally for every thread and never destroyed? The reason I am asking is that we may need a way to do cleanup if they get destroyed frequently.

Reply:
A new state is created every time a new thread processes the rules, and is never destroyed. The number of threads processing rules is not bounded as one might add new frontends (add*Local) or backends (newServer) at runtime, in addition to the ones created at runtime. In most cases the number of threads is quite stable but I guess there might be cases where backends are constantly added and removed at runtime. I'm not sure what to do in that case, though..

Reply:
Are threads ever removed? If they are, one good solution would be to push the Lua state to a pool for later reuse by a new thread.

My main concern is that every state may open a new LMDB env with a large mapsize.

Reply:
Frontend threads are never removed, as we don't support removing a frontend. UDP backend threads do stop when the backend is removed, and thread-local objects should be destroyed automatically when that happens.

Reply:
Assuming the Lua state is also destroyed, I wonder if this would also allow Lua to garbage collect all the objects, in which case there should be no issue.

Reply:
I will test whether this is actually true and will let you know.

Reply:
I confirmed that the Lua state is destroyed when a thread exits, by allocating a global object in the Lua code passed to `LuaFFIPerThreadRule` and making sure the destructor is called after removing a backend. Note that backend-related threads only get response rules and their associated Lua states, not (query) rules.

Reply:
Great, then everything should be fine. Thanks!